File: orb/apps/android/kagami-android/app/src/main/java/com/kagami/android/services/WorldModelService.kt
package com.kagami.android.services

import android.content.Context
import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.withContext
import org.tensorflow.lite.Interpreter
import org.tensorflow.lite.gpu.CompatibilityList
import org.tensorflow.lite.gpu.GpuDelegate
import java.io.FileInputStream
import java.nio.MappedByteBuffer
import java.nio.channels.FileChannel
import javax.inject.Inject
import javax.inject.Singleton

/**
 * World Model inference service for Android using TensorFlow Lite.
 *
 * Runs the Base OrganismRSSM model (50M params) for predictive state modeling.
 *
 * ## Usage
 *
 * ```kotlin
 * val service = WorldModelService(context)
 * service.initialize()
 * val prediction = service.predict(observation, action)
 * ```
 *
 * Created: January 12, 2026
 */
@Singleton
class WorldModelService @Inject constructor(
    private val context: Context
) {
    companion object {
        private const val TAG = "WorldModelService"

        // Model configuration
        private const val MODEL_FILE = "organism_rssm_base.tflite"
        private const val OBS_DIM = 64
        private const val ACTION_DIM = 8
        private const val HIDDEN_DIM = 384
        private const val STOCH_DIM = 32
    }

    // State
    private val _isReady = MutableStateFlow(false)
    val isReady: StateFlow<Boolean> = _isReady.asStateFlow()

    private val _lastLatencyMs = MutableStateFlow(0.0)
    val lastLatencyMs: StateFlow<Double> = _lastLatencyMs.asStateFlow()

    private val _inferenceCount = MutableStateFlow(0)
    val inferenceCount: StateFlow<Int> = _inferenceCount.asStateFlow()

    // TFLite interpreter
    private var interpreter: Interpreter? = null
    private var gpuDelegate: GpuDelegate? = null

    // Cached hidden states
    private var cachedH: FloatArray? = null
    private var cachedZ: FloatArray? = null

    /**
     * World model prediction result.
     */
    data class Prediction(
        val obsPred: FloatArray,
        val reward: Float,
        val continueProb: Float,
        val latencyMs: Double
    ) {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false

            other as Prediction

            if (!obsPred.contentEquals(other.obsPred)) return false
            if (reward != other.reward) return false
            if (continueProb != other.continueProb) return false
            if (latencyMs != other.latencyMs) return false

            return true
        }

        override fun hashCode(): Int {
            var result = obsPred.contentHashCode()
            result = 31 * result + reward.hashCode()
            result = 31 * result + continueProb.hashCode()
            result = 31 * result + latencyMs.hashCode()
            return result
        }
    }

    /**
     * Initialize the world model.
     *
     * Loads the TFLite model from assets with optional GPU acceleration.
     */
    suspend fun initialize(): Result<Unit> = withContext(Dispatchers.IO) {
        try {
            Log.i(TAG, "Initializing world model: $MODEL_FILE")
            val start = System.nanoTime()

            // Load model from assets
            val modelBuffer = loadModelFile()
            if (modelBuffer == null) {
                Log.w(TAG, "World model not found, using placeholder mode")
                _isReady.value = true
                return@withContext Result.success(Unit)
            }

            // Configure interpreter options
            val options = Interpreter.Options().apply {
                setNumThreads(4) // Use multiple CPU threads

                // Try GPU delegate if available
                if (CompatibilityList().isDelegateSupportedOnThisDevice) {
                    gpuDelegate = GpuDelegate()
                    addDelegate(gpuDelegate)
                    Log.i(TAG, "GPU delegate enabled")
                }
            }

            interpreter = Interpreter(modelBuffer, options)

            val elapsed = (System.nanoTime() - start) / 1_000_000.0
            Log.i(TAG, "World model loaded in ${elapsed}ms")

            _isReady.value = true
            Result.success(Unit)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to initialize world model", e)
            Result.failure(e)
        }
    }

    private fun loadModelFile(): MappedByteBuffer? {
        return try {
            val assetFileDescriptor = context.assets.openFd(MODEL_FILE)
            val inputStream = FileInputStream(assetFileDescriptor.fileDescriptor)
            val fileChannel = inputStream.channel
            val startOffset = assetFileDescriptor.startOffset
            val declaredLength = assetFileDescriptor.declaredLength
            fileChannel.map(FileChannel.MapMode.READ_ONLY, startOffset, declaredLength)
        } catch (e: Exception) {
            Log.w(TAG, "Model file not found in assets: $MODEL_FILE")
            null
        }
    }

    /**
     * Predict next state given observation and action.
     *
     * @param observation Current observation vector [OBS_DIM].
     * @param action Action to take [ACTION_DIM].
     * @return Prediction with next state and reward.
     */
    suspend fun predict(observation: FloatArray, action: FloatArray): Result<Prediction> =
        withContext(Dispatchers.Default) {
            try {
                val start = System.nanoTime()

                // If interpreter not loaded, return placeholder
                val interp = interpreter
                if (interp == null) {
                    Log.d(TAG, "World model not loaded, returning placeholder prediction")
                    return@withContext Result.success(
                        Prediction(
                            obsPred = observation.copyOf(),
                            reward = 0f,
                            continueProb = 1f,
                            latencyMs = 0.0
                        )
                    )
                }

                // Prepare inputs
                val obsInput = Array(1) { Array(1) { observation } }
                val actionInput = Array(1) { Array(1) { action } }
                val hInput = Array(1) { cachedH ?: FloatArray(HIDDEN_DIM) }
                val zInput = Array(1) { cachedZ ?: FloatArray(STOCH_DIM) }

                // Prepare outputs
                val obsPredOutput = Array(1) { Array(1) { FloatArray(OBS_DIM) } }
                val hOutput = Array(1) { FloatArray(HIDDEN_DIM) }
                val zOutput = Array(1) { FloatArray(STOCH_DIM) }
                val rewardOutput = Array(1) { FloatArray(1) }
                val continueOutput = Array(1) { FloatArray(1) }

                // Map inputs and outputs
                val inputs = arrayOf(obsInput, actionInput, hInput, zInput)
                val outputs = mapOf(
                    0 to obsPredOutput,
                    1 to hOutput,
                    2 to zOutput,
                    3 to rewardOutput,
                    4 to continueOutput
                )

                // Run inference
                interp.runForMultipleInputsOutputs(inputs, outputs)

                // Update cached states
                cachedH = hOutput[0]
                cachedZ = zOutput[0]

                // Compute continue probability (sigmoid)
                val continueLogit = continueOutput[0][0]
                val continueProb = 1f / (1f + kotlin.math.exp(-continueLogit))

                val elapsed = (System.nanoTime() - start) / 1_000_000.0
                _lastLatencyMs.value = elapsed
                _inferenceCount.value = _inferenceCount.value + 1

                Result.success(
                    Prediction(
                        obsPred = obsPredOutput[0][0],
                        reward = rewardOutput[0][0],
                        continueProb = continueProb,
                        latencyMs = elapsed
                    )
                )
            } catch (e: Exception) {
                Log.e(TAG, "Prediction failed", e)
                Result.failure(e)
            }
        }

    /**
     * Imagine a trajectory given initial state and action sequence.
     *
     * @param initialObs Starting observation.
     * @param actions Sequence of actions.
     * @return Sequence of predictions.
     */
    suspend fun imagine(
        initialObs: FloatArray,
        actions: List<FloatArray>
    ): Result<List<Prediction>> = withContext(Dispatchers.Default) {
        try {
            val predictions = mutableListOf<Prediction>()
            var currentObs = initialObs

            // Reset state for imagination
            resetState()

            for (action in actions) {
                val prediction = predict(currentObs, action).getOrThrow()
                predictions.add(prediction)
                currentObs = prediction.obsPred
            }

            Result.success(predictions)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    /**
     * Reset hidden states (e.g., at episode boundary).
     */
    fun resetState() {
        cachedH = null
        cachedZ = null
    }

    /**
     * Release resources.
     */
    fun close() {
        interpreter?.close()
        interpreter = null

        gpuDelegate?.close()
        gpuDelegate = null

        _isReady.value = false
    }
}

File: orb/apps/ios/kagami-ios/KagamiIOS/Services/WorldModelService.swift
import Foundation
import CoreML
import os.log

/// World Model inference service for iOS using CoreML.
///
/// Runs the Base OrganismRSSM model (50M params) for predictive state modeling.
///
/// # Usage
///
/// ```swift
/// let service = WorldModelService.shared
/// try await service.initialize()
/// let prediction = try await service.predict(observation: obs, action: action)
/// ```
///
/// Created: January 12, 2026
@MainActor
public final class WorldModelService: ObservableObject {

    /// Shared singleton instance.
    public static let shared = WorldModelService()

    // MARK: - Published State

    /// Whether the model is loaded and ready.
    @Published public private(set) var isReady: Bool = false

    /// Last prediction latency in milliseconds.
    @Published public private(set) var lastLatencyMs: Double = 0

    /// Total inference count.
    @Published public private(set) var inferenceCount: Int = 0

    // MARK: - Private Properties

    private let logger = Logger(subsystem: "com.kagami.ios", category: "WorldModel")

    /// CoreML model instance.
    private var model: MLModel?

    /// Cached hidden state.
    private var cachedH: MLMultiArray?

    /// Cached stochastic state.
    private var cachedZ: MLMultiArray?

    /// Model configuration.
    private let config = WorldModelConfig.base

    // MARK: - Configuration

    /// World model configuration.
    public struct WorldModelConfig {
        let modelName: String
        let obsDim: Int
        let actionDim: Int
        let hiddenDim: Int
        let stochDim: Int

        /// Small model for embedded devices.
        public static let small = WorldModelConfig(
            modelName: "organism_rssm_small",
            obsDim: 64,
            actionDim: 8,
            hiddenDim: 256,
            stochDim: 16
        )

        /// Base model for mobile devices.
        public static let base = WorldModelConfig(
            modelName: "organism_rssm_base",
            obsDim: 64,
            actionDim: 8,
            hiddenDim: 384,
            stochDim: 32
        )
    }

    // MARK: - Initialization

    private init() {}

    /// Initialize the world model.
    ///
    /// Loads the CoreML model from the app bundle.
    public func initialize() async throws {
        logger.info("Initializing world model: \(self.config.modelName)")

        let start = CFAbsoluteTimeGetCurrent()

        // Load model from bundle
        guard let modelURL = Bundle.main.url(
            forResource: config.modelName,
            withExtension: "mlmodelc"
        ) else {
            // Try mlpackage
            guard let packageURL = Bundle.main.url(
                forResource: config.modelName,
                withExtension: "mlpackage"
            ) else {
                logger.warning("World model not found in bundle, using placeholder")
                isReady = true
                return
            }

            model = try await loadModel(from: packageURL)
            return
        }

        model = try await loadModel(from: modelURL)

        let elapsed = (CFAbsoluteTimeGetCurrent() - start) * 1000
        logger.info("World model loaded in \(elapsed, privacy: .public)ms")

        isReady = true
    }

    private func loadModel(from url: URL) async throws -> MLModel {
        let configuration = MLModelConfiguration()
        configuration.computeUnits = .all // Use Neural Engine + GPU + CPU

        return try await withCheckedThrowingContinuation { continuation in
            MLModel.load(contentsOf: url, configuration: configuration) { result in
                switch result {
                case .success(let model):
                    continuation.resume(returning: model)
                case .failure(let error):
                    continuation.resume(throwing: error)
                }
            }
        }
    }

    // MARK: - Prediction

    /// World model prediction result.
    public struct Prediction {
        /// Predicted next observation.
        public let obsPred: [Float]

        /// Predicted reward.
        public let reward: Float

        /// Episode continuation probability.
        public let continueProb: Float

        /// Inference latency in milliseconds.
        public let latencyMs: Double
    }

    /// Predict next state given observation and action.
    ///
    /// - Parameters:
    ///   - observation: Current observation vector [obs_dim].
    ///   - action: Action to take [action_dim].
    ///
    /// - Returns: Prediction with next state and reward.
    public func predict(observation: [Float], action: [Float]) async throws -> Prediction {
        let start = CFAbsoluteTimeGetCurrent()

        // If model not loaded, return placeholder prediction
        guard model != nil else {
            logger.debug("World model not loaded, returning placeholder prediction")
            return Prediction(
                obsPred: observation,
                reward: 0.0,
                continueProb: 1.0,
                latencyMs: 0.0
            )
        }

        // Prepare inputs
        let obsArray = try createMultiArray(
            from: observation,
            shape: [1, 1, NSNumber(value: config.obsDim)]
        )

        let actionArray = try createMultiArray(
            from: action,
            shape: [1, 1, NSNumber(value: config.actionDim)]
        )

        // Get or initialize hidden states
        let h = try cachedH ?? createMultiArray(
            from: [Float](repeating: 0, count: config.hiddenDim),
            shape: [1, NSNumber(value: config.hiddenDim)]
        )

        let z = try cachedZ ?? createMultiArray(
            from: [Float](repeating: 0, count: config.stochDim),
            shape: [1, NSNumber(value: config.stochDim)]
        )

        // Create feature provider
        let inputFeatures = try MLDictionaryFeatureProvider(dictionary: [
            "obs": MLFeatureValue(multiArray: obsArray),
            "actions": MLFeatureValue(multiArray: actionArray),
            "h": MLFeatureValue(multiArray: h),
            "z": MLFeatureValue(multiArray: z),
        ])

        // Run inference
        let output = try await model!.prediction(from: inputFeatures)

        // Extract outputs
        let obsPredArray = output.featureValue(for: "obs_pred")?.multiArrayValue
        let obsPred = extractFloats(from: obsPredArray, count: config.obsDim)

        let rewardValue = output.featureValue(for: "reward")?.multiArrayValue
        let reward = rewardValue?[0].floatValue ?? 0.0

        let continueValue = output.featureValue(for: "continue_prob")?.multiArrayValue
        let continueLogit = continueValue?[0].floatValue ?? 0.0
        let continueProb = 1.0 / (1.0 + exp(-continueLogit)) // Sigmoid

        // Update cached states
        if let hOut = output.featureValue(for: "h_out")?.multiArrayValue {
            cachedH = hOut
        }
        if let zOut = output.featureValue(for: "z_out")?.multiArrayValue {
            cachedZ = zOut
        }

        let elapsed = (CFAbsoluteTimeGetCurrent() - start) * 1000
        lastLatencyMs = elapsed
        inferenceCount += 1

        return Prediction(
            obsPred: obsPred,
            reward: reward,
            continueProb: continueProb,
            latencyMs: elapsed
        )
    }

    /// Imagine a trajectory given initial state and action sequence.
    ///
    /// - Parameters:
    ///   - initialObs: Starting observation.
    ///   - actions: Sequence of actions.
    ///
    /// - Returns: Sequence of predictions.
    public func imagine(initialObs: [Float], actions: [[Float]]) async throws -> [Prediction] {
        var predictions: [Prediction] = []
        var currentObs = initialObs

        // Reset state for imagination
        resetState()

        for action in actions {
            let prediction = try await predict(observation: currentObs, action: action)
            predictions.append(prediction)
            currentObs = prediction.obsPred
        }

        return predictions
    }

    /// Reset hidden states (e.g., at episode boundary).
    public func resetState() {
        cachedH = nil
        cachedZ = nil
    }

    // MARK: - Helpers

    private func createMultiArray(from array: [Float], shape: [NSNumber]) throws -> MLMultiArray {
        let multiArray = try MLMultiArray(shape: shape, dataType: .float32)

        for (index, value) in array.enumerated() {
            multiArray[index] = NSNumber(value: value)
        }

        return multiArray
    }

    private func extractFloats(from multiArray: MLMultiArray?, count: Int) -> [Float] {
        guard let array = multiArray else {
            return [Float](repeating: 0, count: count)
        }

        var result: [Float] = []
        for i in 0..<min(count, array.count) {
            result.append(array[i].floatValue)
        }

        // Pad if needed
        while result.count < count {
            result.append(0)
        }

        return result
    }
}

// MARK: - SwiftUI Preview Support

#if DEBUG
extension WorldModelService {
    /// Create a mock service for previews.
    static var preview: WorldModelService {
        let service = WorldModelService()
        service.isReady = true
        return service
    }
}
#endif
